R/combinevariablesetsasbinary.R

```diff
@@ -0,0 +1,161 @@
+        }
+
+        # Consider generalizing in future
+        if (is.null(question.type)) {
+            stop("This function should only be applied to variable sets in Displayr.")
+        }
+            
+
+        if (question.type == "PickOneMulti") {
+            x = x[[1]]
+        }
+        
+        if (is.factor(x)) {
+            levs = levels(x)
+            return(levs[duplicated(levs)])
+        } else {
+            return(colnames(x)[duplicated(colnames(x))])
+        }
+    })
+
+    n.duplicates = vapply(duplicated.labels, FUN = length, FUN.VALUE = numeric(1))
+
+    if (any(n.duplicates > 0)) {
+        dup.qs = names(duplicated.labels)[n.duplicates > 0]
+        dup.labels = duplicated.labels[n.duplicates > 0]
+        stop(paste0("The input data contains duplicate labels and cannot be matched. Duplicated labels: " , dup.labels[[1]]))
+    }
+
+    
+    binary.versions <- lapply(variable.set.list, FUN = questionToBinary)
+
+    # Pick One - Multis will appear as a list.
+    # Append thoses lists to the main list and delete the
+    # corresponding original elements    
+    for (j in 1:length(binary.versions)) {
+        if (!is.data.frame(binary.versions[[j]]) & is.list(binary.versions[[j]])) {
+            binary.versions <- c(binary.versions, binary.versions[[j]])
+            binary.versions[j] <- NULL
+        }
+    }
+    
+    # If only one variable set then just return it
+    if (length(binary.versions) == 1) {
+        result <- binary.versions[[1]] == 1
+    } else {
+
+        # Check matching of column labels in binary data
+        all.labels = lapply(binary.versions, FUN = colnames)
+        unique.labels = unique(unlist(all.labels))
+        common.labels = unique.labels
+        for (j in seq_along(binary.versions)) {
+            common.labels = intersect(common.labels, colnames(binary.versions[[j]]))
+        }
+        if (!setequal(unique.labels, common.labels)) {
+            stop(paste0("Unable to match categories from the input data. The labels which do not appear in all inputs are: ", paste0(setdiff(unique.labels, common.labels), collapse = ",")))
+        }
+
+        input.args = binary.versions
+        input.args[["match.elements"]] <- "Yes"
+        input.args[["elements.to.count"]] <- list(numeric = NA, categorical = NULL)
+        input.args[["ignore.missing"]] <- TRUE
+        
+        # Count missing values for each case for each binary variable
+        n.missing <- do.call(Count, input.args)
+        
+        # Combine the sets of binary variables using AnyOf
+        input.args[["elements.to.count"]] <- list(numeric = 1, categorical = NULL)
+        result <- do.call(AnyOf, input.args)
+        
+        # Handle missing values
+        if (compute.for.incomplete) { # Only assign NA if all missing
+            result[n.missing == length(binary.versions)] <- NA
+        } else { # Assign NA if any missing
+            result[n.missing > 0] <- NA
+        }
+    }
+    result
```

Comment:
Could tidy this slightly by not having the `else` path but putting `return(binary.versions[[1]] == 1)` in the `if` block